src/Sylius/Component/Core/Cart/Provider/NewCartProvider.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Cart\Provider;
+
+use Sylius\Component\Cart\Model\CartInterface;
+use Sylius\Component\Cart\Provider\CartProviderInterface;
+use Sylius\Component\Channel\Context\ChannelNotFoundException;
+use Sylius\Component\Core\Context\ShopperContextInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class NewCartProvider implements CartProviderInterface
```

Comment:
Got another idea, if we'll inject a cart factory here instead of decorating another provider, it can still be named `NewCartProvider`. If not, we should emphasize, that it is a decorator and does all the blaming on passed cart.